### PR TITLE
Fixed: ParseUser.getCurrentUser() returns a non authenticated user when local datastore is used

### PR DESCRIPTION
**Pull request by lukas1994**
Repro project: https://www.dropbox.com/s/m6vl215mg3nj8uo/Parse-Starter-Project-1.9.4.zip
1- change keys
2- click SIGN UP
-> make sure a user & a post are created
3- click GETCURRENTUSER
-> make sure user and session are shown
4- click ISAUTHENTICATED()
-> make sure it returns true
5- now click QUERY DATA WITH INCLUDE("USER")
6- repeat step 3 and 4 and notice the issue

### CODE REVIEW

**Comment by grantland** [comment]
I'm not convinced that this is where the actual problem lies, since we shouldn't be overwriting `sessionToken` when merging from LDS when the JSON isn't complete. Could you add some unit tests that reproduces the problem so I could see more into why this problem occurs?

**Comment by lukas1994** [comment]
If the server returns the query that contains a reference to the ```currentUser``` the following happens:
The ```ParseDecoder``` calls ```ParseObject.fromJSON()``` with ```isComplete == true```.
Then ```ParseObject.createWithoutData()``` gets called.
If the local datastore is enabled it gets ```currentUser``` from the store (by object id).
Finally, we call ```currentUser.mergeFromServer()``` with ```isComplete == true``` which overwrites ```currentUser``` with the data from the server (which doesn't contain the sessionToken, ...).

There are different ways to fix this. We need to add a special case somewhere: ```ParseUser``` is the best place I guess. Another fix would be to set ```isComplete == false``` appropriately.

What do you think?

**Comment by lukas1994** [comment]
One interesting thing to note is that the server normally returns the ```sessionToken``` if you query for users but if you query for posts that contain a pointer to a user then it doesn't return the ```sessionToken```.

**Comment by grantland** [comment]
If that's the case, it seems like this current modification will have some unintended side affects such as stale data such as keys that were removed on the server persisting after a `fetch`.

We'd probably want to have a more precise fix that is conditional upon the `ParseUser` instancing being current and only persisting the `sessionToken` locally if there was no `sessionToken` in the new state.

Additionally, this should be verified with actual code that hits the server, but the PR should contain a unit test that prevents this from happening in the future.

File: Parse/src/main/java/com/parse/ParseUser.java [hunk 1]
```diff
@@ -501,6 +501,15 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
   }
 
   @Override
+  void setState(ParseObject.State newState) {
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ParseUser.java]: For future PRs, we use `/* package */` to designate package protected methods.

File: Parse/src/main/java/com/parse/ParseUser.java [hunk 2]
```diff
@@ -501,6 +501,18 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
...
+  /* package */ void setState(ParseObject.State newState) {
+    // Avoid overwriting session token
+    State.Builder builder = newState.newBuilder();
+    if (isCurrentUser() && getSessionToken() != null &&
+            newState.get("sessionToken") == null) {
+      builder.put("sessionToken", getSessionToken());
+    }
+
+    super.setState(builder.build());
+  }
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ParseUser.java]: It seems like we'd be creating unnecessary `State` objects with every `setState()`. Mind only creating a new `State` if necessary?

File: Parse/src/main/java/com/parse/ParseUser.java [hunk 3]
```diff
@@ -501,6 +501,18 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
   }
 
   @Override
+  /* package */ void setState(ParseObject.State newState) {
+    // Avoid overwriting session token
+    State.Builder builder = newState.newBuilder();
+    if (isCurrentUser() && getSessionToken() != null &&
+            newState.get("sessionToken") == null) {
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ParseUser.java]: nit: multi-line conditionals should have the conditional `&&` on the new line.

File: Parse/src/main/java/com/parse/ParseUser.java [hunk 4]
```diff
@@ -501,6 +501,18 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
   }
 
   @Override
+  /* package */ void setState(ParseObject.State newState) {
+    // Avoid overwriting session token
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ParseUser.java]: Let's add a little more description here. Something like:

    // Avoid clearing sessionToken when updating the current user's State via ParseQuery result

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 5]
```diff
@@ -1059,6 +1061,29 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+  public void testDontOverwriteSessionTokenForCurrentUser() throws Exception {
+    ParseUser.State sessionTokenState = new ParseUser.State.Builder()
+            .sessionToken("sessionToken")
+            .build();
+    ParseUser.State emptyState = new ParseUser.State.Builder().build();
+
+    ParseUser user = ParseObject.from(sessionTokenState);
+    user.setIsCurrentUser(true);
+
+    Assert.assertEquals(user.getSessionToken(), "sessionToken");
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: nit: you should be able to use static imports.

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 6]
```diff
@@ -1059,6 +1061,29 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+
+    user.setState(emptyState);
+
+    Assert.assertEquals(user.getSessionToken(), "sessionToken");
+
+    user.setIsCurrentUser(false);
+
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: nit: whitespace

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 7]
```diff
@@ -1059,6 +1061,29 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+    user.setState(emptyState);
+
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: nit: whitespace

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 8]
```diff
@@ -1059,6 +1061,29 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+            .build();
+    ParseUser.State emptyState = new ParseUser.State.Builder().build();
+
+    ParseUser user = ParseObject.from(sessionTokenState);
+    user.setIsCurrentUser(true);
+
+    Assert.assertEquals(user.getSessionToken(), "sessionToken");
+
+    user.setState(emptyState);
+
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: nit: whitespace

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 9]
```diff
@@ -1059,6 +1061,29 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+  @Test
+  public void testDontOverwriteSessionTokenForCurrentUser() throws Exception {
+    ParseUser.State sessionTokenState = new ParseUser.State.Builder()
+            .sessionToken("sessionToken")
+            .build();
+    ParseUser.State emptyState = new ParseUser.State.Builder().build();
+
+    ParseUser user = ParseObject.from(sessionTokenState);
+    user.setIsCurrentUser(true);
+
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: nit: whitespace

**Comment by grantland** [comment]
This is looking much better! Just an optimization and a few minor nits.

File: Parse/src/main/java/com/parse/ParseUser.java [hunk 10]
```diff
@@ -501,6 +501,18 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
...
   @Override
+  /* package */ void setState(ParseObject.State newState) {
+    // Avoid clearing sessionToken when updating the current user's State via ParseQuery result
+    if (isCurrentUser() && getSessionToken() != null
+            && newState.get("sessionToken") == null) {
+      super.setState(newState.newBuilder().put("sessionToken", getSessionToken()).build());
+    }
+    else {
+      super.setState(newState);
+    }
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ParseUser.java]: This logic seems a tad messy, could we simplify it to the following?

```java
if (/* ... */) {
  newState = // ...
}
super.setState(newState);
```

File: Parse/src/test/java/com/parse/ParseUserTest.java [hunk 11]
```diff
@@ -1059,6 +1059,25 @@ public void testUpgradeToRevocableSessionAsync() throws Exception {
...
+            .sessionToken("sessionToken")
+            .build();
+    ParseUser.State emptyState = new ParseUser.State.Builder().build();
+
+    ParseUser user = ParseObject.from(sessionTokenState);
+    user.setIsCurrentUser(true);
+    assertEquals(user.getSessionToken(), "sessionToken");
+
+    user.setState(emptyState);
+    assertEquals(user.getSessionToken(), "sessionToken");
```
Comment (grantland) [inline Parse/src/test/java/com/parse/ParseUserTest.java]: Could we add some data in `emptyState` that we would want to merge in and verify that it's getting merged in correctly? A key that was in `sessionTokenState` that should be removed, a key that should be overwritten, and a key that should be added?

**Comment by grantland** [comment]
Sorry, I realized that there are some additional use cases that we should cover in the tests.

**Comment by grantland** [comment]
Looks awesome, thanks!

Also same comments RE commits, etc. as I made on your other diff.

**Pull request merged**